### PR TITLE
Update imports/exports to use destructuring properly

### PR DESCRIPTION
**Pull request by samouri**
**summary**
There many files that need to be modified in order to [enable scope hoisting and tree shaking](https://github.com/Automattic/wp-calypso/pull/16057#issuecomment-335047339) and disable the cjs module transform. They need to be modified becaus we have been using es6 modules more loosely than actually [allowed by the spec](https://github.com/Automattic/wp-calypso/pull/16057#issuecomment-335047339). 

The nice thing about stacking most of the fixes into one big PR is that we can test it all by turning off cjs transform and having webpack guarantee that all of our imports are valid

---

This PR fixes all of the issues marked here:  https://github.com/Automattic/wp-calypso/pull/16057#issuecomment-335047339 and is a blocker for disabling commonjs transpilation.

To Review
-----
I recommend viewing the version of the diff [without whitespace](https://github.com/Automattic/wp-calypso/pull/18838/files?w=1).  90% of the changes are just the addition of extra exports.

To Test 
------
first test _without_ cjs transpilation (where we eventually want to get).  the point of this is to ensure that we haven't broken/missed any imports/exports and to make sure everything is still connected together properly.  If we messed up an import, webpack would tell us.  You can see the error reporting behavior locally by messing up an import and you'll see it in massive yellow font.
1. `npm start`.  
2. none of the files should still have **build** errors (runtime errors shall abound)

Then also test _with_ cjs transpilation.  This should work 100%
1. `git checkout update/exports-imports-es6-with-cjs`
2. npm start

### CODE REVIEW

**Comment by samouri** [comment]
this is blocked on fixing the 8 issues found here: 
https://github.com/Automattic/wp-calypso/pull/16057#issuecomment-335757214

**Comment by samouri** [comment]
I've cherry-picked https://github.com/Automattic/wp-calypso/pull/18864/commits/ec6a6c7fece5f93417653d5cf94e5a2e59bbba22 from https://github.com/Automattic/wp-calypso/pull/18864

**Comment by samouri** [comment]
Woot! Here are all of the errors found after turning off cjs transpilation!

```
WARNING in ./client/state/signup/optional-dependencies/reducer.js
14:4-27 "export 'suggestedUsernameSchema' was not found in './schema'

WARNING in ./client/state/notices/middleware.js
214:1724-1752 "export 'PUBLICIZE_CONNECTION_REFRESH' was not found in 'state/action-types'
 
WARNING in ./client/state/notices/middleware.js
214:1812-1848 "export 'PUBLICIZE_CONNECTION_REFRESH_FAILURE' was not found in 'state/action-types'

WARNING in ./node_modules/notifications-panel/src/rest-client/index.js
291:4-16 "export 'default' (imported as 'repliesCache') was not found in '../comment-replies-cache'

WARNING in ./node_modules/notifications-panel/src/Notifications.jsx
29:0-12 "export 'default' (imported as 'repliesCache') was not found in './comment-replies-cache'

WARNING in ./node_modules/notifications-panel/src/templates/comment-reply-input.jsx
31:11-23 "export 'default' (imported as 'repliesCache') was not found in '../comment-replies-cache'

WARNING in ./client/my-sites/sharing/connections/services-group.jsx
44:20-45 "export 'hasOwnProperty' (imported as 'Components') was not found in './services'

WARNING in ./client/state/sharing/publicize/publicize-actions/reducer.js
61:21-43 "export 'publicizeActionsSchema' was not found in './schema'

WARNING in ./client/state/sharing/publicize/publicize-actions/reducer.js
68:4-26 "export 'publicizeActionsSchema' was not found in './schema'

WARNING in ./client/state/site-settings/exporter/reducers.js
97:7-36 "export 'EXPORT_ADVANCED_SETTINGS_FAIL' was not found in 'state/action-types'
```


If you'd like to review this, I recommend [turning off whitespace changes](https://github.com/Automattic/wp-calypso/pull/18838/files?w=1).  It drops the line count down to ~1.2k.

**Comment by samouri** [comment]
I don't want to hold off reviewing this because of merge conflicts.  conflicting files are super easy to fix (checkout new version, add more exports).  So lets review and I'll fix conflicts right before merging

**Comment by psealock** [comment]
Taking this branch for a spin (first without disababling cjs transpilation. See a few errors from Reader's front page, looking into it now:


screen shot 2017-10-20 at 11 00 21 am

**Comment by psealock** [comment]
Fixed the above errors in https://github.com/Automattic/wp-calypso/pull/18838/commits/325df8e0c5c75b53195e7b2476df92a7134a6bdd

`getLocaleSlug` needed to also be included in the default export.

**Comment by psealock** [comment]
Found a site controller bug

screen shot 2017-10-20 at 1 24 35 pm

Fixed in https://github.com/Automattic/wp-calypso/pull/18838/commits/21cdc84d73da7177dd7619308ae4554d4693a866

**Comment by psealock** [comment]
After cherry-picking the two commits, `npm start` produces the following error:

```
ERROR in ./client/boot/app.js
'import' and 'export' may only appear at the top level (5:0)
You may need an appropriate loader to handle this file type.
```

**Comment by samouri** [comment]
> After cherry-picking the two commits, npm start produces the following error:

We discussed in slack and the issue was caused by only having the first commit and not the second.

I pushed both relevant commits to this branch temporarily: 
1. https://github.com/Automattic/wp-calypso/pull/16057/commits/20084651071cbdc0544e29626ab6104076e52f38 - disable commonjs transpilation
2. https://github.com/Automattic/wp-calypso/pull/16057/commits/50fa7a8663d11fac21e522322465c01ab9b161ff - disable add-module-exports plugin

The plan is to _not merge these two commits_.  They are just useful to make webpack check that all modules are being imported/exported properly at build time.  Calypso will fail at runtime due to all of the requires

File: webpack.config.js [hunk 1]
```diff
@@ -20,6 +20,10 @@ const AssetsPlugin = require( 'assets-webpack-plugin' );
 const cacheIdentifier = require( './server/bundler/babel/babel-loader-cache-identifier' );
 const config = require( './server/config' );
 const UseMinifiedFiles = require( './server/bundler/webpack-plugins/use-minified-files' );
+const babelConfig = JSON.parse( fs.readFileSync( './.babelrc', { encoding: 'utf8' } ) );
+
+babelConfig.presets[ 0 ][ 1 ].modules = false;
+babelConfig.plugins = _.without( babelConfig.plugins, 'add-module-exports' );
```
Comment (gziolo) [inline webpack.config.js]: Is it introduced to keep the existing setup for server files and tests? 🤔

Comment (samouri) [reply webpack.config.js]: Yep. And also the bin files that generate component statistics etc. More than a couple pieces rely on the current babelrc file, so I thought this was the cleanest for now.

Comment (samouri) [reply webpack.config.js]: note that i am reverting this before merging. its just for people to have an easy time verifying w/ webpack that all of the imports/exports are configured properly.

Comment (gziolo) [reply webpack.config.js]: Yes, it all makes sense 👍

**Comment by dmsnell** [comment]
For reference, `phone-input/data.js` appears to be [generated](https://github.com/Automattic/wp-calypso/blob/b2038a9c7170a53af8463fa92786b3e1aa8e0771/bin/build-metadata.js)

**Comment by dmsnell** [comment]
@samouri I rebased this against the latest **master** - previous two commits were (your) a39eacc6ac69accbb35a406dcdb63b5d70b75cf9 and (mine) b3bb21c78be4cdb31c2cb632ffa2398fa87b4f38

Hope I didn't screw anything up in the process. I verified manually through until the post-feed-store files and then had to stop.

**Comment by samouri** [comment]
@dmsnell: nice! thanks for updating this.  
We can verify that no errors were introduced by running it without the cjs transform and checking for compile time errors. I'll do that Monday

**Comment by dmsnell** [comment]
@samouri I squashed and rebased again; former HEAD was 84a7f27

Did this to try and keep the PR moving and up-to-date, not to rewrite your history or intentionally lose old commits.

File: client/lib/feed-post-store/actions.js [hunk 2]
```diff
@@ -102,4 +104,7 @@ FeedPostActions = {
 	},
 };
 
+const { fetchPost, receivePost, markSeen } = FeedPostActions;
+export { fetchPost, receivePost, markSeen };
```
Comment (dmsnell) [inline client/lib/feed-post-store/actions.js]: why this and not…

```js
export const { fetchPost, receivePost, markSeen } = FeedPostActions;
```

is this a result of the codemod not recognizing that they could be combined?

**Comment by dmsnell** [comment]
This won't seem to run for me so there's a chance I messed up the PR. I'm going to try and cut it up into some smaller pieces…

File: client/lib/feed-post-store/actions.js (hunk 2 above)
Comment (samouri) [reply client/lib/feed-post-store/actions.js]: i don't remember if it was as a result of me or a codemod at this point. its been awhile.

**Comment by samouri** [comment]
Closed because they were all invidivuallly solved in separate prs in the [esmodules milestone](https://github.com/Automattic/wp-calypso/milestone/224)

**Comment by samouri** [comment]
Closed because they were all invidivuallly solved in separate prs in the [esmodules milestone](https://github.com/Automattic/wp-calypso/milestone/224)

**Pull request closed without merging**